Not too helpful termination error messages

The following program produces error messages that could be more helpful.

```Haskell
module Test where
{-@ LIQUID "--ple" @-}

{-@ reflect ms @-}
{-@ ms :: [[Int]] -> Nat @-}
ms :: [[Int]] -> Int
ms [] = 0
ms (x:xs) = 1 + m x + ms xs

{-@ reflect m @-}
{-@ m :: [Int] -> Nat @-}
m :: [Int] -> Int
m [] = 0
m (_:xs) = 1 + m xs

{-@ f :: xs:[[Int]] -> Int / [ms xs] @-}
f :: [[Int]] -> Int
f [] = 0
f ([]:xs) = 1 + f xs
f ((_:x):xs) = 1 + f (x:xs)
```

Here are the errors.
<details>

```
Test.hs:24:17: error:
    Termination Check
Property
  {vvRec : _ | vvRec <=> Test.ms xs##aGZ >= 0
                         && Test.ms xs##aGZ < Test.ms ?a}
.
Not provable in context
  ?a := {?a : [[GHC.Internal.Types.Int]] | GHC.Types_LHAssumptions.len ?a >= 0}

  ?b := {?b : [GHC.Internal.Types.Int] | GHC.Types_LHAssumptions.len ?b >= 0}

  ?c := {?c : [[GHC.Internal.Types.Int]] | ?c == ?a
                                           && ?c == GHC.Internal.Types.: ?b xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?c == 1 + GHC.Types_LHAssumptions.len xs##aGZ
                                           && head ?c == ?b
                                           && lqdc##$select##GHC.Internal.Types.:##1 ?c == ?b
                                           && lqdc##$select##GHC.Internal.Types.:##2 ?c == xs##aGZ
                                           && tail ?c == xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?c >= 0}

  ?d := {?d : [GHC.Internal.Types.Int] | ?d == ?b
                                         && ?d == GHC.Internal.Types.[]
                                         && GHC.Types_LHAssumptions.len ?d == 0
                                         && GHC.Types_LHAssumptions.len ?d >= 0}

  xs##aGZ := {v : [[GHC.Internal.Types.Int]] | GHC.Types_LHAssumptions.len v >= 0}
   |
24 | f ([]:xs) = 1 + f xs
   |                 ^^^^



Test.hs:25:20: error:
    Termination Check
Property
  {vvRec : _ | vvRec <=> Test.ms ?f >= 0
                         && Test.ms ?f < Test.ms ?a}
.
Not provable in context
  ?a := {?a : [[GHC.Internal.Types.Int]] | GHC.Types_LHAssumptions.len ?a >= 0}

  ?b := {?b : [GHC.Internal.Types.Int] | GHC.Types_LHAssumptions.len ?b >= 0}

  ?c := GHC.Internal.Types.Int

  ?d := {?d : [[GHC.Internal.Types.Int]] | ?d == ?a
                                           && ?d == GHC.Internal.Types.: ?b xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?d == 1 + GHC.Types_LHAssumptions.len xs##aGZ
                                           && head ?d == ?b
                                           && lqdc##$select##GHC.Internal.Types.:##1 ?d == ?b
                                           && lqdc##$select##GHC.Internal.Types.:##2 ?d == xs##aGZ
                                           && tail ?d == xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?d >= 0}

  ?e := {?e : [GHC.Internal.Types.Int] | ?e == ?b
                                         && ?e == GHC.Internal.Types.: ?c x##aH0
                                         && GHC.Types_LHAssumptions.len ?e == 1 + GHC.Types_LHAssumptions.len x##aH0
                                         && head ?e == ?c
                                         && lqdc##$select##GHC.Internal.Types.:##1 ?e == ?c
                                         && lqdc##$select##GHC.Internal.Types.:##2 ?e == x##aH0
                                         && tail ?e == x##aH0
                                         && GHC.Types_LHAssumptions.len ?e >= 0}

  ?f := {?f : [[GHC.Internal.Types.Int]] | ?f == GHC.Internal.Types.: x##aH0 xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?f == 1 + GHC.Types_LHAssumptions.len xs##aGZ
                                           && head ?f == x##aH0
                                           && lqdc##$select##GHC.Internal.Types.:##1 ?f == x##aH0
                                           && lqdc##$select##GHC.Internal.Types.:##2 ?f == xs##aGZ
                                           && tail ?f == xs##aGZ
                                           && GHC.Types_LHAssumptions.len ?f >= 0}

  x##aH0 := {v : [GHC.Internal.Types.Int] | GHC.Types_LHAssumptions.len v >= 0}

  xs##aGZ := {v : [[GHC.Internal.Types.Int]] | GHC.Types_LHAssumptions.len v >= 0}
   |
25 | f ((_:x):xs) = 1 + f (x:xs)
   | 
```
</details>

Ideally, we would get an indication of the identifier of the constraint that is failing, so we can check the ple dumps, for instance.

For those curious, termination doesn't pass here because we are using `reflect` instead of `measure`. `measure` would make the fact that the metrics return natural numbers more readily available to the SMT solver.

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/Split.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/Split.hs
@@ -276,7 +276,7 @@ splitC _ (SubR γ o r)
     r2  = F.RR F.boolSort $ F.Reft (vv, F.EVar vv)
     vv  = "vvRec"
     ci  = Ci src err (cgVar γ)
-    err = Just $ ErrAssType src o (text $ show o ++ "type error") g (rHole rr)
+    err = Just $ ErrAssType src o (text $ show o ++ "type error") Nothing g (rHole rr)
     rr  = toReft r
     tag = getTag γ
     src = getLocation γ

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Liquid.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Liquid.hs
@@ -161,6 +161,7 @@ cidE2u cfg (subcId, e) =
   where
     attachSubcId es@ErrSubType{}      = es { cid = Just subcId }
     attachSubcId es@ErrSubTypeModel{} = es { cid = Just subcId }
+    attachSubcId es@ErrAssType{}      = es { cid = Just subcId }
     attachSubcId es = es
 
 -- writeCGI tgt cgi = {-# SCC "ConsWrite" #-} writeFile (extFileName Cgi tgt) str

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Types/Errors.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Types/Errors.hs
@@ -259,6 +259,7 @@ data TError t =
   | ErrAssType { pos  :: !SrcSpan
                , obl  :: !Oblig
                , msg  :: !Doc
+               , cid  :: Maybe SubcId
                , ctx  :: !(M.HashMap Symbol t)
                , cond :: t
                } -- ^ condition failure error
@@ -801,10 +802,11 @@ hint e = maybe empty (\d -> "" $+$ ("HINT:" <+> d)) (go e)
 --------------------------------------------------------------------------------
 ppError' :: (PPrint a, Show a) => Tidy -> Doc -> TError a -> Doc
 --------------------------------------------------------------------------------
-ppError' td dCtx (ErrAssType _ o _ c p)
+ppError' td dCtx (ErrAssType _ o _ cid c p)
   = pprintTidy td o
         $+$ dCtx
         $+$ ppFull td (ppPropInContext td p c)
+        $+$ maybe mempty (\i -> text "Constraint id" <+> text (show i)) cid
 
 ppError' td dCtx err@(ErrSubType _ _ _ _ _ tE)
   | totalityType td tE

